src/cpu.c

```diff
@@ -21,55 +22,70 @@ cpu* mgec_new_cpu() {
   return c;
 }
 
-#define load(C, R, V) _Generic((V), u8: load8, u16: load16)(C, R, V)
-
-static inline cycles load8(cpu* c, cpu_register_t reg, u8 value) {
+inline cycles load8(cpu* c, cpu_register_t reg, u8 value) {
+  // TODO: flag changes
+  // zero out the high byte when in native mode or
+  // emulation mode with 8-bit indexers
+  bool hbo = (c->e || (c->sr_x || c->sr_m));
   switch (reg) {
   case XL:
     c->xl = value;
-    break;
-  case XH:
-    c->xh = value;
+    if (hbo)
+      c->xh = 0;
     break;
   case YL:
     c->yl = value;
+    if (hbo)
+      c->yh = 0;
     break;
-  case YH:
-    c->yh = value;
-    break;
-  // Making sure we're handling all cases
   // TODO: some error condition
   case X:
+  case XH:
   case Y:
+  case YH:
   case A:
   case B:
   case C:
   default:
     break;
   }
 
+  if (value == 0) {
+    c->sr_z = 1;
+  } else {
+    c->sr_z = 0;
+  }
+
   return 1;
 }
 
-static inline cycles load16(cpu* c, cpu_register_t reg, u16 value) {
+inline cycles load16(cpu* c, cpu_register_t reg, u16 value) {
+  // TODO: flag changes
   switch (reg) {
   case X:
     c->x = value;
     break;
   case Y:
     c->y = value;
     break;
+  case C:
+    c->c = value;
   // TODO: some error condition
   case XL:
   case XH:
   case YL:
   case YH:
   case A:
   case B:
-  case C:
   default:
     break;
   }
 
-  return 1;
-}
\ No newline at end of file
+  if (value == 0) {
```

Comment:
Looks good here, but we're also missing the negative check here, which in this case should be bit 15 (still the most significant bit).